### ISSUE TITLE
kaminari pagination always selected in favor of will_paginate

### ISSUE THREAD

**Issue by gato-omega** [elastic/elasticsearch-rails#423]
I've just noticed this, when having both kaminari and will_paginate included in a project, elasticsearch-rails selects kaminari always in favor of will_paginate.

This prioritization occurs at this line:

https://github.com/elastic/elasticsearch-rails/blob/v0.1.7/elasticsearch-model/lib/elasticsearch/model.rb#L38

I've opened this as an issue due to problems when paginating elasticsearch results without actually having the pagination to work, and without any errors being thrown either. The actual problem we had was caused with RailsAdmin requiring Kaminari and the pagination for elasticsearch results not working because we were trying to use will_paginate's methods (but no errors were thrown).

in order to solve the issue at the moment, as a workaround, one can include the will_paginate module which will effectively solve the issue but will leave some pollution with the kaminari methods still there.

``` ruby
# add this to solve, make sure you load this code after elasticsearch-rails loads
Elasticsearch::Model::Response::Response.__send__ :include, Elasticsearch::Model::Response::Pagination::WillPaginate
```

Of course, the Kaminari methods can be undefined by using remove_method or undef_method cycling through the already included module and such... but as a cleaner solution, giving the opportunity to choose the pagination library (at setup/config time) could help in avoiding these hard to track problems (or even, not requiring any integration by default, so that the developer is forced to make a conscious decision).

Thanks to @Johaned for taking some time to track down this with me. Hope this issue helps overcome pagination not working (and not throwing errors) in some scenarios. I read about a possible related issue here: https://github.com/elastic/elasticsearch-rails/issues/239.

I am filing this so that any future developers (including myself!) with this same strange issue can have a workaround (if they ever encounter this).

Thanks!

**Comment by miguelff** [comment]
Hey @gato-omega Thanks for reporting this, and for the workaround.

I was aware of this, but left it out of the scope of the implementation, cause I thought it would be a very edge case to have two pagination gems used for different models.

Can I ask you a question, to better guess how desirable will fixing this?

Besides a temporary migration, which will be a use-case for having two different pagination gems under the same project, used in different models?

Thank you!

**Comment by miguelff** [comment]
Comment updated!

**Comment by gato-omega** [comment]
Thank you for your quick reply!

Ah yes, it is just because we were using RailsAdmin (sorry, not ActiveAdmin, I updated my comment above too) and Kaminari was being required as a dependency without us explicitly mentioning it in the Gemfile.

I would say this is really not very critical, although we did not know why it was not working until tracking down it was Kaminari that somehow got into the project.

In some sense we really did not want to use two pagination libraries at the same time, it just happened that we installed will_paginate without knowing Kaminari was already there because of another gem and elasticsearch pagination was not working because it automatically used kaminari's integration, the real issue here is that we did not get any errors or anything.

Long story short, the desirability of this fix could really not be so high, instead people may find the workaround if they ever need it, since I also did not see anyone else with this same problem. More likely people will appreciate more the fact that the library automatically integrates with their pagination library, and removing this feature would introduce a breaking change, so... yeah.

**Comment by gato-omega** [comment]
Now that I think twice, perhaps it is really good to have the option to strictly specify the pagination library at some point in the future, just in case people are giving up in using will_paginate whenever this happens and instead they just decide to use kaminari instead.

I mentioned that this would probably introduce a breaking change, but not quite, perhaps a good solution which is backwards compatible is to throw an error whenever both libraries are detected, letting know the developer what to do, and this would still work with previous codebases that only have one pagination library installed across the project.

Cheers!

**Comment by celesteking** [comment]
How do I use WillPaginate with ES Persistence gem?
